Anti-replay mechanism (#92)
ietf-wg-ohai/oblivious-http: draft-ietf-ohai-ohttp.md

# HPKE Encapsulation

HTTP message encapsulation uses HPKE for request and response encryption.

> An encapsulated HTTP request includes a timestamp and a binary-encoded HTTP

Reviewer: Does the encapsulated request actually include a timestamp? The format doesn't seem to. If this is just referring to the fact that there should be a Date header, it seems misleading.

Author: Editing error on my part; I missed this one when I cribbed from #88.

~~~
{: #fig-enc-request title="Encapsulated Request"}

> The Nenc parameter corresponding to the HpkeKdfId can be found in {{!HPKE}}.

Reviewer 2: ```suggestion
The Nenc parameter corresponding to the HpkeKdfId can be found in {{!HPKE, Section 7.1}}.
```

Author: ```suggestion
The Nenc parameter corresponding to the HpkeKdfId can be found in {{Section 7.1 of !HPKE}}.
```

successfully decrypted using the key.

A server MUST ensure that the HPKE keys it uses are not valid for any other
protocol that uses HPKE with the "message/bhttp request" label.  Designers of

> other protocols, especially new versions of this protocol, can ensure key

Reviewer 2: ```suggestion
other protocols reusing this encapsulation format, especially new versions of this protocol, can ensure key
```

Author: ```suggestion
protocols that reuse this encapsulation format, especially new versions of this protocol, can ensure key
```

diversity by choosing a different label in their use of HPKE.  The
"message/bhttp response" label was chosen for symmetry only as it provides key
diversity only within the HPKE context created using the "message/bhttp request"

> label.

Reviewer 2: ```suggestion
label. See {{repurposing-the-encapsulation-format}} for more details.
```

Author: ```suggestion
label; see {{repurposing-the-encapsulation-format}}.
```

> A server is responsible for either rejecting replayed request or ensuring that

Reviewer 2: ```suggestion
A server is responsible for either rejecting replayed requests or ensuring that
```

risk, though it is important to recognize that different idempotent requests
can be combined to be not idempotent.

Idempotent actions with a narrow scope based on the value of an
application-provided nonce could enable data submission with limited replay
protection.

Even without strong replay prevention, the server-chosen `response_nonce` field
ensures that responses have unique AEAD keys and nonces even when requests are
replayed.

### Use of Date for Anti-Replay

Clients SHOULD include a `Date` header field in requests.  Though HTTP requests
often do not include a `Date` header field, the value of this field might be
used by a server to limit the amount of requests it needs to track when
preventing replays.

A server can maintain state for requests for a small window of time over which
it wishes to accept requests.  The server then rejects requests if the request
is the same as one that was previously answered within that time window.
Servers can identify duplicate requests using the `enc` value.  The server can

> reject requests if the Date request header ffield is outside of the chosen time

Reviewer: `ffield` -> `field`

risk, though it is important to recognize that different idempotent requests
can be combined to be not idempotent.

Even without strong replay prevention, the server-chosen `response_nonce` field
ensures that responses have unique AEAD keys and nonces even when requests are
replayed.

### Use of Date for Anti-Replay

> Clients SHOULD include a `Date` header field in requests.  Though HTTP requests

Reviewer 2: ```suggestion
Clients SHOULD include a `Date` header field in encapsulated requests.  Though HTTP requests
```

often do not include a `Date` header field, the value of this field might be
used by a server to limit the amount of requests it needs to track when
preventing replays.

A server can maintain state for requests for a small window of time over which
it wishes to accept requests.  The server then rejects requests if the request

> is the same as one that was previously answered within that time window.

Author: If the new request has a timestamp in the new time window, it can accept it as a new value.

Generally you want a sliding window for anti-replay, so that you have a window of W = (now() - X .. now() + Y) continuously.  Then, when a request comes in with t \in W, you accept it if the enc value is unique.  You drop old values when t < now() - X. 

In practice, this isn't very efficient and you need to quantize the time window a little bit.  The anti-replay I built for TLS 0-RTT uses just two time intervals.  A single window means that there are hard discontinuities that cause rejections for very small amounts of skew when you move from one window to the next.

Either way, once you forget the old enc value, if you see the same enc value with a new value for t, it's a new request and you can accept it.

> is the same as one that was previously answered within that time window.

Author: ```suggestion
is the same as one that was previously answered within that time window.  Servers can reject
requests outside of this window and signal that clients might retry with a different `Date`
header field; see {{Section 4 of !REQUEST-DATE}}.
```

> Servers can identify duplicate requests using the `enc` value.  The server can

Reviewer 2: ```suggestion
Servers can identify duplicate requests using the encapsulation `enc` value.  The server can
```

risk, though it is important to recognize that different idempotent requests
can be combined to be not idempotent.

Even without replay prevention, the server-chosen `response_nonce` field
ensures that responses have unique AEAD keys and nonces even when requests are
replayed.

### Use of Date for Anti-Replay

> Clients SHOULD include a `Date` header field in encapsulated requests.  Though HTTP requests

Reviewer: Should we call out that this is only when using binary http as the inner protocol?

Reviewer: It seems like we're missing the bigger normative requirement that "that the client and target server SHOULD ensure that the protocol that’s encapsulated isn’t going to be vulnerable to replay attacks". That doesn't seem to be stated anywhere, and that's really the main point.

Author: Good point.  I've added some text at the top of the anti-replay section (second paragraph).  Let me know if that makes sense.

Reviewer: @martinthomson Yeah, that's exactly the kind of text I was thinking of. Thanks for adding!
